File: py/src/t_plat/config/loader.py
"""The layered loader: defaults → files → environment → secret resolution.

Mirrors ``rust/crates/platform-config/src/loader.rs``.
"""

from __future__ import annotations

import os
import tomllib
from collections.abc import Mapping
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from t_plat.config.errors import ConfigFileError
from t_plat.config.merge import apply_env_overrides, deep_merge
from t_plat.config.model import Config, Profile
from t_plat.config.secret import (
    OsSecretSource,
    ResolvedSecrets,
    SecretSource,
)

__all__ = [
    "BASE_FILE_NAME",
    "DEFAULT_CONFIG_DIR",
    "ENV_CONFIG_DIR",
    "ENV_CONFIG_FILE",
    "ENV_PROFILE",
    "ConfigLoader",
    "LayerSource",
    "LoadedConfig",
    "load_config",
]

#: Environment variable selecting the profile.
ENV_PROFILE = "T_PLAT_PROFILE"
#: Environment variable selecting the directory holding the config files.
ENV_CONFIG_DIR = "T_PLAT_CONFIG_DIR"
#: Environment variable naming an extra file layered above the profile file.
ENV_CONFIG_FILE = "T_PLAT_CONFIG_FILE"
#: Directory searched for ``default.toml`` / ``<profile>.toml`` when
#: :data:`ENV_CONFIG_DIR` is unset. Relative to the working directory.
DEFAULT_CONFIG_DIR = "config"
#: Filename of the base layer inside the config directory.
BASE_FILE_NAME = "default.toml"


@dataclass(frozen=True, slots=True)
class LayerSource:
    """Where one contributing layer came from.

    Recorded in load order (lowest precedence first) so the effective config
    can be explained.
    """

    kind: str
    """``"builtin"``, ``"file"`` or ``"env"``."""

    detail: str
    """The file path, or the comma-separated keys the environment overrode."""

    @classmethod
    def builtin(cls) -> LayerSource:
        """Layer 1: :meth:`Config.defaults`."""
        return cls("builtin", "Config.defaults()")

    @classmethod
    def file(cls, path: Path) -> LayerSource:
        """Layers 2–4: a TOML file that existed and was merged."""
        return cls("file", str(path))

    @classmethod
    def env(cls, keys: list[str]) -> LayerSource:
        """Layer 5: ``T_PLAT__*`` overrides, with the keys they changed."""
        return cls("env", ", ".join(keys))


@dataclass(frozen=True, slots=True)
class LoadedConfig:
    """The result of a successful load."""

    config: Config
    """The typed configuration. Contains secret *references*, never values."""

    secrets: ResolvedSecrets
    """The secret values resolved for this config, keyed by dotted config key."""

    sources: list[LayerSource]
    """The layers that contributed, lowest precedence first."""


@dataclass(frozen=True, slots=True)
class ConfigLoader:
    """Builds a :class:`LoadedConfig` from the documented precedence chain.

    The environment and the secret source are both injected rather than read
    from the process, so tests can cover precedence and the ``prod`` secret
    policy hermetically.
    """

    env: Mapping[str, str] = field(default_factory=dict)
    """The environment snapshot to read control variables and overrides from."""

    profile: Profile | None = None
    """Forces the profile, ignoring :data:`ENV_PROFILE`."""

    config_dir: Path | None = None
    """Forces the config directory, ignoring :data:`ENV_CONFIG_DIR`."""

    config_file: Path | None = None
    """Forces the explicit file layer, ignoring :data:`ENV_CONFIG_FILE`."""

    secret_source: SecretSource = field(default_factory=OsSecretSource)
    """Where secret material is fetched from."""

    def __repr__(self) -> str:
        # The env snapshot may hold secret material; print only its shape.
        return (
            f"ConfigLoader(env_vars={len(self.env)}, profile={self.profile!r}, "
            f"config_dir={self.config_dir!r}, config_file={self.config_file!r})"
        )

    def load(self) -> LoadedConfig:
        """Run the full precedence chain and resolve secrets.

        Raises a :class:`~t_plat.config.errors.ConfigError` subclass on an
        unreadable or malformed file, an invalid profile, an unknown or
        unparseable environment override, a schema violation (including a
        literal secret), an out-of-range value, a profile rule violation, or —
        under ``prod`` — an unresolvable secret.
        """
        profile = self._resolve_profile()
        config_dir = self._resolve_config_dir()
        sources = [LayerSource.builtin()]

        # Layer 1: built-in defaults. Guarantees every schema key exists, so
        # environment overrides always have a declared type to parse against.
        document = Config.defaults().to_table()

        # Layers 2 and 3: base file, then the profile overlay. Both optional —
        # the built-in defaults are a complete config on their own.
        for candidate in (config_dir / BASE_FILE_NAME, config_dir / f"{profile}.toml"):
            if candidate.is_file():
                document = deep_merge(document, _read_toml(candidate))
                sources.append(LayerSource.file(candidate))

        # Layer 4: the explicit file. Requested explicitly, so it must exist.
        explicit = self._resolve_config_file()
        if explicit is not None:
            if not explicit.is_file():
                raise ConfigFileError(
                    f"config file `{explicit}` was requested explicitly but does not exist"
                )
            document = deep_merge(document, _read_toml(explicit))
            sources.append(LayerSource.file(explicit))

        # Layer 5: typed environment overrides beat every file.
        overridden = apply_env_overrides(document, self.env)
        if overridden:
            sources.append(LayerSource.env(overridden))

        config = Config.from_table(document, profile)
        config.validate()
        config.validate_profile_rules()

        # Secrets live outside the merge chain: the layers only ever carried
        # references, and this is where they are dereferenced.
        policy = profile.secret_policy
        entries = {}
        for key, reference in config.secret_refs():
            secret = reference.resolve(key, self.secret_source, policy)
            if secret is not None:
                entries[key] = secret

        return LoadedConfig(config=config, secrets=ResolvedSecrets(entries), sources=sources)

    def _resolve_profile(self) -> Profile:
        """Builder override, else :data:`ENV_PROFILE`, else ``local``."""
        if self.profile is not None:
            return self.profile
        raw = self.env.get(ENV_PROFILE)
        return Profile.LOCAL if raw is None else Profile.parse(raw)

    def _resolve_config_dir(self) -> Path:
        """Builder override, else :data:`ENV_CONFIG_DIR`, else ``./config``."""
        if self.config_dir is not None:
            return self.config_dir
        return Path(self.env.get(ENV_CONFIG_DIR, DEFAULT_CONFIG_DIR))

    def _resolve_config_file(self) -> Path | None:
        """Builder override, else :data:`ENV_CONFIG_FILE`, else no extra layer."""
        if self.config_file is not None:
            return self.config_file
        raw = self.env.get(ENV_CONFIG_FILE)
        return None if raw is None else Path(raw)


def _read_toml(path: Path) -> dict[str, Any]:
    """Read and parse a TOML file that must exist."""
    try:
        text = path.read_text(encoding="utf-8")
    except OSError as exc:
        raise ConfigFileError(f"cannot read config file `{path}`: {exc}") from exc
    try:
        return tomllib.loads(text)
    except tomllib.TOMLDecodeError as exc:
        raise ConfigFileError(f"cannot parse config file `{path}`: {exc}") from exc


def load_config(
    *,
    env: Mapping[str, str] | None = None,
    profile: Profile | None = None,
    config_dir: Path | str | None = None,
    config_file: Path | str | None = None,
    secret_source: SecretSource | None = None,
) -> LoadedConfig:
    """Load configuration using the documented precedence chain.

    With no arguments this reads the real process environment and ``./config``.
    Every argument is an override for testing or embedding.
    """
    return ConfigLoader(
        env=os.environ if env is None else env,
        profile=profile,
        config_dir=None if config_dir is None else Path(config_dir),
        config_file=None if config_file is None else Path(config_file),
        secret_source=OsSecretSource() if secret_source is None else secret_source,
    ).load()

File: rust/crates/platform-config/src/loader.rs
//! The layered loader: defaults → files → environment → secret resolution.

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

use toml::Value;

use crate::error::ConfigError;
use crate::merge::{apply_env_overrides, merge_into};
use crate::model::{Config, Profile};
use crate::secret::{OsSecretSource, ResolvedSecrets, SecretSource};

/// Environment variable selecting the profile.
pub const ENV_PROFILE: &str = "T_PLAT_PROFILE";
/// Environment variable selecting the directory holding the config files.
pub const ENV_CONFIG_DIR: &str = "T_PLAT_CONFIG_DIR";
/// Environment variable naming an extra file layered above the profile file.
pub const ENV_CONFIG_FILE: &str = "T_PLAT_CONFIG_FILE";
/// Directory searched for `default.toml` / `<profile>.toml` when
/// [`ENV_CONFIG_DIR`] is unset. Relative to the process working directory.
pub const DEFAULT_CONFIG_DIR: &str = "config";
/// Filename of the base layer inside the config directory.
pub const BASE_FILE_NAME: &str = "default.toml";

/// Where one contributing layer came from. Recorded in load order (lowest
/// precedence first) so the effective config can be explained.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum LayerSource {
    /// Layer 1: [`Config::default`].
    BuiltinDefaults,
    /// Layers 2–4: a TOML file that existed and was merged.
    File(PathBuf),
    /// Layer 5: `T_PLAT__*` overrides, with the dotted keys they changed.
    Environment(Vec<String>),
}

/// The result of a successful load.
#[derive(Clone, Debug)]
pub struct LoadedConfig {
    /// The typed configuration. Contains secret *references*, never values.
    pub config: Config,
    /// The secret values resolved for this config, keyed by dotted config key.
    pub secrets: ResolvedSecrets,
    /// The layers that contributed, lowest precedence first.
    pub sources: Vec<LayerSource>,
}

/// Builds a [`LoadedConfig`] from the documented precedence chain.
///
/// The environment and the secret source are both injected rather than read
/// from the process, so tests can cover precedence and the `prod` secret
/// policy hermetically — see the crate-level docs for an example.
pub struct Loader {
    env: BTreeMap<String, String>,
    profile: Option<Profile>,
    config_dir: Option<PathBuf>,
    config_file: Option<PathBuf>,
    secret_source: Box<dyn SecretSource>,
}

impl std::fmt::Debug for Loader {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        // The env snapshot may hold secret material; print only its shape.
        f.debug_struct("Loader")
            .field("env_vars", &self.env.len())
            .field("profile", &self.profile)
            .field("config_dir", &self.config_dir)
            .field("config_file", &self.config_file)
            .finish_non_exhaustive()
    }
}

impl Loader {
    /// Build a loader from an explicit environment snapshot and an in-memory
    /// secret source (i.e. nothing is read from the process).
    #[must_use]
    pub fn from_env_map(env: BTreeMap<String, String>) -> Self {
        Self {
            env,
            profile: None,
            config_dir: None,
            config_file: None,
            secret_source: Box::new(OsSecretSource),
        }
    }

    /// Build a loader from the real process environment.
    #[must_use]
    pub fn from_os_env() -> Self {
        Self::from_env_map(std::env::vars().collect())
    }

    /// Force the profile, ignoring [`ENV_PROFILE`].
    #[must_use]
    pub fn profile(mut self, profile: Profile) -> Self {
        self.profile = Some(profile);
        self
    }

    /// Force the config directory, ignoring [`ENV_CONFIG_DIR`].
    #[must_use]
    pub fn config_dir(mut self, dir: impl Into<PathBuf>) -> Self {
        self.config_dir = Some(dir.into());
        self
    }

    /// Force the explicit config file layer, ignoring [`ENV_CONFIG_FILE`].
    ///
    /// The file must exist; an explicit request is never silently skipped.
    #[must_use]
    pub fn config_file(mut self, file: impl Into<PathBuf>) -> Self {
        self.config_file = Some(file.into());
        self
    }

    /// Resolve secrets against `source` instead of the process environment and
    /// filesystem.
    #[must_use]
    pub fn secret_source(mut self, source: impl SecretSource + 'static) -> Self {
        self.secret_source = Box::new(source);
        self
    }

    /// Run the full precedence chain and resolve secrets.
    ///
    /// # Errors
    ///
    /// Any [`ConfigError`]: an unreadable or malformed file, an invalid
    /// profile, an unknown or unparseable environment override, a schema
    /// violation (including a literal secret), an out-of-range value, a
    /// profile rule violation, or — under `prod` — an unresolvable secret.
    pub fn load(&self) -> Result<LoadedConfig, ConfigError> {
        let profile = self.resolve_profile()?;
        let config_dir = self.resolve_config_dir();
        let mut sources = vec![LayerSource::BuiltinDefaults];

        // Layer 1: built-in defaults. Guarantees every schema key exists, so
        // environment overrides always have a declared type to parse against.
        let mut document = builtin_defaults_document()?;

        // Layers 2 and 3: base file, then the profile overlay. Both optional —
        // the built-in defaults are a complete config on their own.
        for candidate in [
            config_dir.join(BASE_FILE_NAME),
            config_dir.join(format!("{profile}.toml")),
        ] {
            if let Some(layer) = read_optional_toml(&candidate)? {
                merge_into(&mut document, layer);
                sources.push(LayerSource::File(candidate));
            }
        }

        // Layer 4: the explicit file. Requested explicitly, so it must exist.
        if let Some(path) = self.resolve_config_file() {
            if !path.is_file() {
                return Err(ConfigError::MissingExplicitFile { path });
            }
            merge_into(&mut document, read_toml(&path)?);
            sources.push(LayerSource::File(path));
        }

        // Layer 5: typed environment overrides beat every file.
        let overridden = apply_env_overrides(&mut document, &self.env)?;
        if !overridden.is_empty() {
            sources.push(LayerSource::Environment(overridden));
        }

        let mut config = deserialize(&document)?;
        config.profile = profile;
        config.validate()?;
        config.validate_profile_rules()?;

        // Secrets live outside the merge chain: the layers only ever carried
        // references, and this is where they are dereferenced.
        let policy = profile.secret_policy();
        let mut secrets = ResolvedSecrets::default();
        for (key, reference) in config.secret_refs() {
            if let Some(secret) = reference.resolve(key, self.secret_source.as_ref(), policy)? {
                secrets.insert(key, secret);
            }
        }

        Ok(LoadedConfig {
            config,
            secrets,
            sources,
        })
    }

    /// Builder override, else [`ENV_PROFILE`], else [`Profile::Local`].
    fn resolve_profile(&self) -> Result<Profile, ConfigError> {
        if let Some(profile) = self.profile {
            return Ok(profile);
        }
        match self.env.get(ENV_PROFILE) {
            Some(raw) => raw.parse(),
            None => Ok(Profile::default()),
        }
    }

    /// Builder override, else [`ENV_CONFIG_DIR`], else [`DEFAULT_CONFIG_DIR`].
    fn resolve_config_dir(&self) -> PathBuf {
        self.config_dir
            .clone()
            .or_else(|| self.env.get(ENV_CONFIG_DIR).map(PathBuf::from))
            .unwrap_or_else(|| PathBuf::from(DEFAULT_CONFIG_DIR))
    }

    /// Builder override, else [`ENV_CONFIG_FILE`], else no explicit layer.
    fn resolve_config_file(&self) -> Option<PathBuf> {
        self.config_file
            .clone()
            .or_else(|| self.env.get(ENV_CONFIG_FILE).map(PathBuf::from))
    }
}

/// Layer 1 as a TOML document.
fn builtin_defaults_document() -> Result<Value, ConfigError> {
    let rendered = toml::to_string(&Config::default()).map_err(|err| ConfigError::Schema {
        message: format!("built-in defaults are not serializable: {err}"),
    })?;
    toml::from_str::<Value>(&rendered).map_err(|err| ConfigError::Schema {
        message: format!("built-in defaults are not valid TOML: {err}"),
    })
}

/// Read and parse a TOML file that must exist.
fn read_toml(path: &Path) -> Result<Value, ConfigError> {
    let text = std::fs::read_to_string(path).map_err(|source| ConfigError::ReadFile {
        path: path.to_path_buf(),
        source,
    })?;
    toml::from_str::<Value>(&text).map_err(|source| ConfigError::ParseFile {
        path: path.to_path_buf(),
        source,
    })
}

/// Read and parse a TOML file, or `Ok(None)` if it is simply absent.
fn read_optional_toml(path: &Path) -> Result<Option<Value>, ConfigError> {
    if path.is_file() {
        read_toml(path).map(Some)
    } else {
        Ok(None)
    }
}

/// Turn the merged document into the typed schema.
///
/// Goes back through text so the deserializer's error message names the
/// offending key (which is how a literal secret gets reported usefully).
fn deserialize(document: &Value) -> Result<Config, ConfigError> {
    let rendered = toml::to_string(document).map_err(|err| ConfigError::Schema {
        message: format!("merged config is not serializable: {err}"),
    })?;
    toml::from_str::<Config>(&rendered).map_err(|err| ConfigError::Schema {
        message: err.to_string(),
    })
}
